### PR TITLE
Index search terms

### PR DESCRIPTION
**Pull request by rashkov**
To review, I'm building up an index of (search term -> [list of articles which the term appears in])
My approach is to take each article, get all of its words, then map each word to a (word, article_id) pair. Then I'm doing groupByKey on the word to get the final (word, [list of articles]), and then doing mapValues(distinct) to de-duplicate [list of articles]. 

Unfortunately, the groupByKey causes java.lang.OutOfMemoryError: GC Overhead Limit Exceeded.

I'm looking into ways to tune this job. The general advice would be to avoid shuffles by partitioning the data correctly, but I don't think that approach works for this particular data. Any advice would be appreciated, thank you!

### CODE REVIEW

**Comment by aravindr18** [comment]
Hi @rashkov, have you read this article: https://databricks.gitbooks.io/databricks-spark-knowledge-base/content/best_practices/prefer_reducebykey_over_groupbykey.html? 
TLDR; Avoid using groupByKey whenever possible. 
Here are a few things that you could try: 
- Try doubling the memory of executors to see if the problem persists 
- See if you can accomplish what you're trying to do using the `collect_list` method  in spark 

Let me know if that helps

**Comment by aravindr18** [comment]
A simple example would be as follows: 
Assuming in the source df you have the word and document column, you could do as follows: 
```df.groupBy("word").agg(
  collect_list(col("document")) as "documents"```
Let me know if this helps

File: src/main/scala/graphlytic.scala [hunk 1]
```diff
@@ -1,48 +1,41 @@
...
+  //   val r = new RedisClient("10.0.0.8", 6379)
+  //   r.set("somekey", "someval");
+  //   println(r.get("somekey"))
+  // }
+
+  def main(args: Array[String]) {
+    // setup the Spark Context
+    val conf = new SparkConf()
+      .setAppName(appName)
+      .setMaster("local[2]")
```
Comment (aravindr18) [inline src/main/scala/graphlytic.scala]: @rashkov `setMaster` needs to be removed from here. This is specified as part of your spark-submit.
